File: cases/022.thermostat-sensor-decoupling/src/thermostat_controller.cc
#include "thermostat_controller.h"

namespace nitr::case022 {

ThermostatController::ThermostatController(float target_temperature)
    : target_temperature_(target_temperature) {}

ThermostatController::Command ThermostatController::Evaluate(
    float current_temperature) const {
  if (current_temperature <= target_temperature_ - kTemperatureDeltaThreshold) {
    return Command::kHeating;
  }
  if (current_temperature >= target_temperature_ + kTemperatureDeltaThreshold) {
    return Command::kCooling;
  }
  return Command::kIdle;
}

float ThermostatController::target_temperature() const {
  return target_temperature_;
}

}  // namespace nitr::case022

File: cases/029.thermostat-sensor-decoupling-python/src/thermostat_controller.py
from __future__ import annotations

from enum import Enum


class Command(Enum):
    HEATING = "Heating"
    COOLING = "Cooling"
    IDLE = "Idle"


class ThermostatController:
    TEMPERATURE_DELTA_THRESHOLD = 2.0

    def __init__(self, target_temperature: float) -> None:
        self._target_temperature = target_temperature

    def evaluate(self, current_temperature: float) -> Command:
        if current_temperature <= (
            self._target_temperature - self.TEMPERATURE_DELTA_THRESHOLD
        ):
            return Command.HEATING
        if current_temperature >= (
            self._target_temperature + self.TEMPERATURE_DELTA_THRESHOLD
        ):
            return Command.COOLING
        return Command.IDLE

    def target_temperature(self) -> float:
        return self._target_temperature
